tools/pkgchk/cmd/root.go

```diff
@@ -0,0 +1,254 @@
+	}
+	return exceps, nil
+}
+
+type pkg struct {
+	// the directory where the package resides relative to the root dir
+	d string
+
+	// the AST of the package
+	p *ast.Package
+}
+
+// returns true if the package directory corresponds to an ARM package
+func (p pkg) isARMPkg() bool {
+	return strings.Index(p.d, "/mgmt/") > -1
+}
+
+// walks the directory hierarchy from the specified root returning a slice of all the packages found
+func getPkgs(rootDir string) ([]pkg, error) {
+	pkgs := []pkg{}
+	err := filepath.Walk(rootDir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if info.IsDir() {
+			// check if leaf dir
+			fi, err := ioutil.ReadDir(path)
+			if err != nil {
+				return err
+			}
+			hasSubDirs := false
+			for _, f := range fi {
+				if f.IsDir() {
+					hasSubDirs = true
+					break
+				}
+			}
+			if !hasSubDirs {
+				fs := token.NewFileSet()
+				packages, err := parser.ParseDir(fs, path, nil, parser.PackageClauseOnly)
+				if err != nil {
+					return err
+				}
+				if len(packages) < 1 {
+					return errors.New("didn't find any packages which is unexpected")
+				}
+				if len(packages) > 1 {
+					return errors.New("found more than one package which is unexpected")
+				}
+				var p *ast.Package
+				for _, pkgs := range packages {
+					p = pkgs
+				}
+				// normalize directory separator to '/' character
+				pkgs = append(pkgs, pkg{
+					d: strings.Replace(path[len(rootDir):], "\\", "/", -1),
+					p: p,
+				})
+			}
+		}
+		return nil
+	})
+	return pkgs, err
+}
+
+type verifier func(p pkg) error
+
+// returns a list of verifiers to execute
+func getVerifiers() []verifier {
+	return []verifier{
+		verifyPkgName(),
+		verifyLowerCase(),
+		verifyDirectorySturcture(),
+	}
+}
+
+// ensures that the leaf directory name matches the package name
+func verifyPkgName() verifier {
+	return func(p pkg) error {
+		leaf := p.d[strings.LastIndex(p.d, "/")+1:]
+		if strings.Compare(leaf, p.p.Name) != 0 {
+			return fmt.Errorf("leaf directory of '%s' doesn't match package name '%s'", p.d, p.p.Name)
+		}
+		return nil
+	}
+}
+
+// ensures that there are no upper-case letters in a package's directory
+func verifyLowerCase() verifier {
+	return func(p pkg) error {
+		// walk the package directory looking for upper-case characters
+		for _, r := range p.d {
+			if r == '/' {
+				continue
+			}
+			if unicode.IsUpper(r) {
+				return fmt.Errorf("found upper-case character in directory '%s'", p.d)
+			}
+		}
+		return nil
+	}
+}
+
+// ensures that the package's directory hierarchy is properly formed
+func verifyDirectorySturcture() verifier {
+	return func(p pkg) error {
+		// for ARM the package directory structure is highly deterministic:
+		// /redis/mgmt/2015-08-01/redis
+		// /resources/mgmt/2017-06-01-preview/policy
+		// /preview/signalr/mgmt/2018-03-01-preview/signalr
+		if !p.isARMPkg() {
+			return nil
+		}
+		regexStr := strings.Join([]string{
+			`^(?:/preview)?`,
```

Comment:
is this regex tested ?

Reply:
by that I mean the examples above are exhaustive  ?

Reply:
for ARM yes and the majority of packages match this pattern.  there are just a handful of exceptions which is why I added the exceptions.txt mechanism.